thread.c

```diff
@@ -1189,13 +1188,8 @@ thread_join_sleep(VALUE arg)
         end = rb_hrtime_add(*limit, rb_hrtime_now());
     }
 
-    while (!thread_finished(target_th)) {
-        VALUE scheduler = rb_fiber_scheduler_current();
-
-        if (scheduler != Qnil) {
-            rb_fiber_scheduler_block(scheduler, target_th->self, p->timeout);
-        }
-        else if (!limit) {
+    while (target_th->status != THREAD_KILLED) {
```

Comment:
As long as the GVL prevents the threads in `thread_join_sleep` from continuing until `th->status = THREAD_KILLED`, this is safe. However, I'm not sure this is always true, as `rb_threadptr_join_list_wakeup` could release the GVL.

```
    rb_threadptr_join_list_wakeup(th);
    rb_threadptr_unlock_all_locking_mutexes(th);

    if (th->invoke_type == thread_invoke_type_ractor_proc) {
        rb_thread_terminate_all(th);
        rb_ractor_teardown(th->ec);
    }

    th->status = THREAD_KILLED;
    thread_debug("thread end: %p\n", (void *)th);
```